File: 15.3Sum.py
# https://leetcode.com/problems/3sum/description/
#
# algorithms
# Medium (22.1%)
# Total Accepted:    379.1K
# Total Submissions: 1.7M
# beats 28.09% of python submissions


class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        length = len(nums)
        if length < 3:
            return []
        nums.sort()
        res, i = [], 0

        while i < length:
            while i > 0 and i < length and nums[i] == nums[i - 1]:
                i += 1
            head, tail = i + 1, length - 1
            while head < tail:
                cnt_sum = nums[i] + nums[head] + nums[tail]
                if cnt_sum > 0:
                    tail -= 1
                elif cnt_sum < 0:
                    head += 1
                else:
                    res.append([nums[i], nums[head], nums[tail]])
                    head, tail = head + 1, tail - 1
                    while head < tail and nums[head] == nums[head - 1]:
                        head += 1
                    while tail > head and nums[tail] == nums[tail + 1]:
                        tail -= 1
            i += 1

        return res

File: 15.3Sum.java
// https://leetcode.com/problems/3sum/
//
// algorithms
// Medium (23.97%)
// Total Accepted: 546,654
// Total Submissions: 2,280,298


class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);

        ArrayList<List<Integer>> res = new ArrayList<>();
        int len = nums.length, i = 0;

        while (i < len) {
            int j = i + 1, k = len - 1, target = -nums[i];
            while (j < k) {
                int sum = nums[j] + nums[k];
                if (sum == target) {
                    res.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    j++;
                    k--;
                    while (j < k && nums[j] == nums[j - 1]) {
                        j++;
                    }
                    while (j < k && nums[k] == nums[k + 1]) {
                        k--;
                    }
                } else if (sum > target) {
                    k--;
                } else {
                    j++;
                }
            }
            i++;
            while (i < len && nums[i] == nums[i - 1]) {
                i++;
            }
        }

        return res;
    }
}